proposals/2020-07-01_aws-kms-keyring-redesign/change.md

```diff
@@ -0,0 +1,327 @@
+   to be included in all AWS KMS calls.
+
+**Decryption Contract**
+
+1. The keyring MUST have access to an AWS KMS client that is configured with credentials for an AWS principal
+   that passes an [AWS KMS Decrypt API](https://docs.aws.amazon.com/kms/latest/APIReference/API_Decrypt.html)
+   authorization check for the encrypted data key.
+2. The keyring is unable to write a decryption contract,
+   but it MAY fulfill any AWS KMS decryption contract for the configured AWS region and, if provided, the AWS account ID.
+   1. Fulfilling any AWS KMS decryption contract MUST be sufficient to succeed.
+
+**Initialization**
+
+On keyring initialization, an AWS KMS symmetric region discovery keyring MUST define the following:
+
+- AWS SDK client (and the AWS SDK client’s region)
+
+On keyring initialization, an AWS KMS symmetric region discovery keyring MAY define the following:
+
+- AWS account ID
+- Grant Tokens
+
+### Keyring-Producing Operations
+
+**Please note the following names are not finalized.**
+
+The ESDK MUST provide customers a way to use the following:
+
+1. **AWS KMS symmetric multi-CMK keyring**
+   1. MUST be configured with a list of key names
+   2. MUST be configured with a key name identifying the [generator](https://github.com/awslabs/aws-encryption-sdk-specification/blob/dbc17f93100667e28dc54e64d05a625db3e5bac2/framework/kms-keyring.md#generator)
+   3. MAY be configured with a list of string [grant tokens](https://github.com/awslabs/aws-encryption-sdk-specification/blob/dbc17f93100667e28dc54e64d05a625db3e5bac2/framework/kms-keyring.md#grant-tokens)
+      to be included in all AWS KMS calls
+   4. MUST configure a multi-keyring of AWS KMS symmetric keyrings...
+      1. with a single AWS KMS symmetric keyring for each configured key name
+      2. with each AWS KMS symmetric keyring configured with the configured grant tokens, if applicable
+      3. where the order of the child keyrings matches the order of the configured key names
+2. **AWS KMS symmetric multi-region discovery keyring**
+   1. MUST be configured with a list of AWS regions
+   2. MAY be configured with an AWS account ID
+   3. MAY be configured with a list of string [grant tokens](https://github.com/awslabs/aws-encryption-sdk-specification/blob/dbc17f93100667e28dc54e64d05a625db3e5bac2/framework/kms-keyring.md#grant-tokens)
+      to be included in all AWS KMS calls
+   4. MUST configure a multi-keyring of AWS KMS symmetric region discovery keyrings...
+      1. with a single AWS KMS symmetric region discovery keyring for each configured AWS region
+      2. with each AWS KMS symmetric region discovery keyring configured with the configured AWS account ID, if applicable
+      3. with each AWS KMS symmetric region discovery keyring configured with the configured grant tokens, if applicable
+      4. where the order of the child keyrings matches the order of the configured AWS regions
+
+**All keyring-producing operations MUST...**
```

Comment:
Consider making this something like "The ESDK MUST provide keyring-producing operations for these derived keyrings. All such operations MUST..." Here you really are talking about the operations rather than the end result. All of these last sections should be qualified to refer only to these derived cases, assuming we define "keyring-producing operations" to also apply to initializing the keyring base cases.

Reply:
I updated the text above the derived keyrings to read "The ESDK MUST provide keyring-producing operations for the following derived keyrings:"

I think leaving "**All keyring-producing operations MUST...**" as-is works best. We...
1. Define what keyring-producing operations are/do earlier in the doc
2. Say which ones we need
3. Say "All keyring-producing operations MUST/MAY/MUST NOT"